### PR TITLE
Add BigQuery plugin

### PR DESCRIPTION
**Pull request by kanterov**
# TL;DR
Implements WebAPI plugin for BigQuery.

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [x] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Adds implementation of token exchange mechanism to get Google credentials using workload identity. Implement BigQuery plugin, for now, only to create QUERY jobs, COPY/EXTRACT/LOAD jobs are going to be parts of subsequent pull requests.

- [BigQuery API](https://cloud.google.com/bigquery/docs/reference/rest)
- [Using Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity)
- [Accessing resources from an OIDC identity provider](https://cloud.google.com/iam/docs/access-resources-oidc)

## Tracking Issue
https://github.com/flyteorg/flyte/issues/817

## Follow-up issue
_NA_

### CODE REVIEW

File: go/tasks/plugins/webapi/bigquery/query_job.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+	Location  string `json:"location"`
+	ProjectID string `json:"projectId"`
+
+	// AllowLargeResults: [Optional] If true and query uses legacy SQL
+	// dialect, allows the query to produce arbitrarily large result tables
+	// at a slight cost in performance. Requires destinationTable to be set.
+	// For standard SQL queries, this flag is ignored and large results are
+	// always allowed. However, you must still set destinationTable when
+	// result size exceeds the allowed maximum response size.
+	AllowLargeResults bool `json:"allowLargeResults,omitempty"`
```
Comment (kumare3) [inline go/tasks/plugins/webapi/bigquery/query_job.go:25]: should this be input to the execution or task definition?

Comment (kumare3) [reply go/tasks/plugins/webapi/bigquery/query_job.go:25]: ohh nevermind, i see you are using this as a json input

File: go/tasks/pluginmachinery/google/gke_token_source.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+const (
+	defaultGracePeriod             = 300
+	gcpServiceAccountAnnotationKey = "iam.gke.io/gcp-service-account"
+	workflowIdentityDocURL         = "https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity"
+)
+
+type gkeTokenSource struct {
+	kubeClient             kubernetes.Interface
+	tokens                 *sync.Map
```
Comment (kumare3) [inline go/tasks/pluginmachinery/google/gke_token_source.go:29]: why not use one of the ttl caches we have an implementation for already?

File: go/tasks/pluginmachinery/google/gke_token_source.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return oauth2.StaticTokenSource(value.(*oauth2.Token)), nil
+}
+
+func getKubeClient(kubeConfigPath string, kubeConfig KubeClientConfig) (*kubernetes.Clientset, error) {
```
Comment (kumare3) [inline go/tasks/pluginmachinery/google/gke_token_source.go:164]: you are using the webAPI, why do you need a kube client?

File: go/tasks/pluginmachinery/google/gke_token_source.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+	var kubecfg *rest.Config
+	var err error
+	if kubeConfigPath != "" {
+		kubeConfigPath := os.ExpandEnv(kubeConfigPath)
+		kubecfg, err = clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error building kubeconfig")
+		}
+	} else {
+		kubecfg, err = rest.InClusterConfig()
```
Comment (kumare3) [inline go/tasks/pluginmachinery/google/gke_token_source.go:174]: if you want to do incluster config, better to rely on the kubeclient, that is part of the TaskExecutionContext?

File: go/tasks/plugins/webapi/bigquery/query_job.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+	UseLegacySQL *bool `json:"useLegacySql,omitempty"`
+
+	// UseQueryCache: [Optional] Whether to look for the result in the query
+	// cache. The query cache is a best-effort cache that will be flushed
+	// whenever tables in the query are modified. Moreover, the query cache
+	// is only available when a query does not have a destination table
+	// specified. The default value is true.
+	//
+	// Default: true
+	UseQueryCache *bool `json:"useQueryCache,omitempty"`
```
Comment (kumare3) [inline go/tasks/plugins/webapi/bigquery/query_job.go:127]: weird to have default true, with a pointer? default for golang is `false` for a bool. and with pointer its like a ternary value

File: go/tasks/pluginmachinery/google/gke_token_source.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+func (m *gkeTokenSource) getGcpServiceAccount(ctx context.Context, identity Identity) (string, error) {
+	serviceAccounts := m.kubeClient.CoreV1().ServiceAccounts(identity.K8sNamespace)
+
+	serviceAccountResponse, err := serviceAccounts.Get(ctx, identity.K8sServiceAccount, metav1.GetOptions{})
+
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create k8s service account token")
+	}
+
+	for key, value := range serviceAccountResponse.Annotations {
```
Comment (kanterov) [inline go/tasks/pluginmachinery/google/gke_token_source.go:102]:
```
	if value, ok := serviceAccountResponse.Annotations[gcpServiceAccountAnnotationKey]; ok {
		return value, nil
	}
```

File: go/tasks/pluginmachinery/google/integration_test.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+			if auth != "Bearer federated-access-token" {
+				writer.WriteHeader(401)
+				return
+			}
+
+			writer.WriteHeader(200)
+			response := iamcredentials.GenerateAccessTokenResponse{AccessToken: "google-access-token"}
+			bytes, _ := json.Marshal(response)
+			_, _ = writer.Write(bytes)
```
Comment (kanterov) [inline go/tasks/pluginmachinery/google/integration_test.go:98]: ioutils.WriteAll

File: go/tasks/pluginmachinery/google/integration_test.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+					"owner":                          "abc",
+					"iam.gke.io/gcp-service-account": "gcp-service-account",
+				},
+			},
+		}, nil
+	})
+
+	ctx := context.TODO()
+
+	t.Run("", func(t *testing.T) {
```
Comment (kanterov) [inline go/tasks/pluginmachinery/google/integration_test.go:54]: drop that

File: go/tasks/pluginmachinery/google/sts.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "can't get iss from subject token")
+	}
+
+	if iss == "" {
+		return nil, errors.New("issued federated token doesn't have 'iss' in payload")
+	}
+
+	audience := fmt.Sprintf("identitynamespace:%s:%s", request.IdentityNamespace, iss)
```
Comment (kanterov) [inline go/tasks/pluginmachinery/google/sts.go:101]: https://github.com/istio/istio/blob/master/security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go#L164

File: go/tasks/pluginmachinery/google/sts.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+	service, err := iamcredentials.NewService(ctx, options...)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to construct iamcredentials service")
+	}
+
+	name := "projects/-/serviceAccounts/" + stsRequest.ServiceAccount
+	request := iamcredentials.GenerateAccessTokenRequest{
+		Scope:    stsRequest.Scope,
+		Lifetime: lifetime,
```
Comment (kanterov) [inline go/tasks/pluginmachinery/google/sts.go:233]: move to config

File: go/tasks/pluginmachinery/google/sts.go [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+			apiError, ok := err.(*googleapi.Error)
+
+			if ok {
+				logger.Errorf(ctx, "failed to generate access token: %v", apiError)
+
+				if apiError.Code >= http.StatusBadRequest && apiError.Code < http.StatusInternalServerError {
+					break
+				}
+			} else {
+				break
```
Comment (kanterov) [inline go/tasks/pluginmachinery/google/sts.go:251]: Should retry as well

File: go/tasks/plugins/webapi/bigquery/plugin.go [hunk 12]
```diff
@@ -0,0 +1,503 @@
...
+
+	taskInfo := createTaskInfo(resourceMeta)
+
+	if resource.CreateError != nil {
+		return handleCreateError(resource.CreateError, taskInfo), nil
+	}
+
+	switch resource.Status.State {
+	case "PENDING":
+		return core.PhaseInfoRunning(version, taskInfo), nil
```
Comment (kanterov) [inline go/tasks/plugins/webapi/bigquery/plugin.go:260]: `PhaseInfoInitializing`

File: go/tasks/plugins/webapi/bigquery/plugin.go [hunk 13]
```diff
@@ -0,0 +1,503 @@
...
+
+func createTaskInfo(resourceMeta *ResourceMetaWrapper) *core.TaskInfo {
+	timeNow := time.Now()
+	j := fmt.Sprintf("bq:%s:%s", resourceMeta.JobReference.Location, resourceMeta.JobReference.JobId)
+
+	return &core.TaskInfo{
+		OccurredAt: &timeNow,
+		Logs: []*flyteIdlCore.TaskLog{
+			{
+				Uri: fmt.Sprintf("https://console.cloud.google.com/bigquery?project=%v&j=%v&page=queryresults",
```
Comment (EngHabu) [inline go/tasks/plugins/webapi/bigquery/plugin.go:435]: Mind moving this to a const on top?

File: go/tasks/plugins/webapi/bigquery/plugin.go [hunk 14]
```diff
@@ -0,0 +1,503 @@
...
+		return pluginsCore.PhaseInfoFailed(pluginsCore.PhasePermanentFailure, userExecutionError, taskInfo)
+
+	default:
+		return pluginsCore.PhaseInfoFailed(pluginsCore.PhasePermanentFailure, systemExecutionError, taskInfo)
+	}
+}
+
+func createTaskInfo(resourceMeta *ResourceMetaWrapper) *core.TaskInfo {
+	timeNow := time.Now()
+	j := fmt.Sprintf("bq:%s:%s", resourceMeta.JobReference.Location, resourceMeta.JobReference.JobId)
```
Comment (EngHabu) [inline go/tasks/plugins/webapi/bigquery/plugin.go:429]: Can you also move this to a function with a name representing what it's used for

File: go/tasks/plugins/webapi/bigquery/plugin.go [hunk 15]
```diff
@@ -0,0 +1,503 @@
...
+
+func formatJobReference(reference bigquery.JobReference) string {
+	return fmt.Sprintf("%s:%s.%s", reference.ProjectId, reference.Location, reference.JobId)
+}
+
+func (p Plugin) newBigQueryClient(ctx context.Context, identity google.Identity) (*bigquery.Service, error) {
+	options := []option.ClientOption{
+		option.WithScopes("https://www.googleapis.com/auth/bigquery"),
+		// FIXME how do I access current version?
+		option.WithUserAgent(fmt.Sprintf("%s/%s", "flytepropeller", "LATEST")),
```
Comment (EngHabu) [inline go/tasks/plugins/webapi/bigquery/plugin.go:452]: We should probably put this in the context...

File: go/tasks/plugins/webapi/bigquery/plugin.go [hunk 16]
```diff
@@ -0,0 +1,503 @@
...
+	}
+
+	if p.cfg.bigQueryEndpoint != "" {
```
Comment (EngHabu) [inline go/tasks/plugins/webapi/bigquery/plugin.go:455]:
```suggestion
        // for testing/mocking purposes
	if p.cfg.bigQueryEndpoint != "" {
```

File: go/tasks/pluginmachinery/google/gke_token_source.go (hunk 2 above)
Comment (kanterov) [reply go/tasks/pluginmachinery/google/gke_token_source.go:29]: can you please point out to TTL cache? I've only found auto refresh cache

**Comment by kanterov** [comment]
Didn't have much time to work on it. I'm going through the comments. There is an issue with changing to using Kubernetes client from controller-runtime. It doesn't support sub-resources, and TokenRequest is a sub-resource to ServiceAccount. See https://github.com/kubernetes-sigs/controller-runtime/issues/172.

We have REST client that was originally used on top-level initialization code in flytepropeller, so we can potentially propagate it all the way through, or we can keep it as is. As for now, the second option seems more feasible not to bloat this pull request. 

@EngHabu @kumare3 what do you think?

File: go/tasks/pluginmachinery/google/gke_token_source.go (hunk 2 above)
Comment (kumare3) [reply go/tasks/pluginmachinery/google/gke_token_source.go:29]: ohh here is an example using one - https://github.com/flyteorg/flytepropeller/blob/master/pkg/controller/nodes/task/barrier.go#L28

We just use the k8s apimachinery cache

File: go/tasks/pluginmachinery/google/gke_token_source.go (hunk 3 above)
Comment (kumare3) [reply go/tasks/pluginmachinery/google/gke_token_source.go:166]: i understand now, for the token stuff. We should pass this in in the task context

File: go/tasks/pluginmachinery/google/gke_token_source.go (hunk 2 above)
Comment (kanterov) [reply go/tasks/pluginmachinery/google/gke_token_source.go:29]: Thanks! This is what I was looking for

File: go/tasks/pluginmachinery/google/gke_token_source.go (hunk 3 above)
Comment (kanterov) [reply go/tasks/pluginmachinery/google/gke_token_source.go:166]: @kumare3 have you seen my last comment? this code uses the REST client, and not the client we have in the task context. Not sure what to do about it

File: go/tasks/plugins/webapi/bigquery/config_test.go [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+func TestUnmarshalBigQueryQueryConfig(t *testing.T) {
+	custom := structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			"projectId": structpb.NewStringValue("project-id"),
+			"location":  structpb.NewStringValue("EU"),
+			"query":     structpb.NewStringValue("SELECT 1"),
+		},
+	}
+
+	config, err := unmarshalQueryJobConfig(&custom)
```
Comment (fediazgon) [inline go/tasks/plugins/webapi/bigquery/config_test.go:19]: You can import 

```
"github.com/flyteorg/flyteplugins/go/tasks/pluginmachinery/utils"
```

And use: 

```
queryJobConfig = QueryJobConfig{}
utils.UnmarshalStructToObj(custom, &queryJobConfig)
```

**Comment by kumare3** [comment]
@kanterov Would love to help push this over the boundary. How should we do this?

File: go/tasks/plugins/webapi/bigquery/plugin.go [hunk 18]
```diff
@@ -0,0 +1,503 @@
...
+		Kind:    flyteIdlCore.ExecutionError_SYSTEM,
+		Code:    code,
+	}
+
+	if createError.Code >= http.StatusBadRequest && createError.Code < http.StatusInternalServerError {
+		return core.PhaseInfoFailed(pluginsCore.PhasePermanentFailure, userExecutionError, taskInfo)
+	}
+
+	if createError.Code >= http.StatusInternalServerError {
+		return core.PhaseInfoFailed(pluginsCore.PhaseRetryableFailure, systemExecutionError, taskInfo)
```
[the remark this answers is not in the document]
Comment (kanterov) [reply go/tasks/plugins/webapi/bigquery/plugin.go:302]: I can change, but there is no similar method for non-retryable failures, that loses symmetry in the code and makes it less readable (in my opinion).

File: go/tasks/plugins/webapi/bigquery/config_test.go (hunk 17 above)
Comment (kanterov) [reply go/tasks/plugins/webapi/bigquery/config_test.go:19]: Fixed

**Comment by kanterov** [comment]
@kumare3 @EngHabu as we discussed, I've removed token exchange part (that needs k8s rest client), and addressed comments for the rest of the code.

File: go/tasks/plugins/webapi/bigquery/query_job.go (hunk 1 above)
Comment (kumare3) [reply go/tasks/plugins/webapi/bigquery/query_job.go:25]: do we want to model this as a protobuf (plugins) so that we can get the protobufs built automatically?

File: boilerplate/flyte/golang_test_targets/Makefile [hunk 19]
```diff
@@ -10,7 +10,7 @@ download_tooling: #download dependencies (including test deps) for the package
 
 .PHONY: lint
 lint: download_tooling #lints the package for common code smells
-	GL_DEBUG=linters_output,env golangci-lint run --deadline=5m --exclude deprecated -v
+	GL_DEBUG=linters_output,env golangci-lint run --deadline=7m --exclude deprecated -v
```
Comment (kumare3) [inline boilerplate/flyte/golang_test_targets/Makefile:13]: this needs an update in boilerplate and it will automatically percolate. 
But do you need this?

File: go/tasks/plugins/webapi/bigquery/config.go [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+				Value: 50,
+			},
+		},
+		GoogleTokenSource: google.GetDefaultConfig(),
+	}
+
+	configSection = pluginsConfig.MustRegisterSubSection("bigquery", &defaultConfig)
+)
+
+type Config struct {
```
Comment (kumare3) [inline go/tasks/plugins/webapi/bigquery/config.go:53]: documentation for the config?

File: go/tasks/plugins/webapi/bigquery/plugin.go [hunk 21]
```diff
@@ -0,0 +1,503 @@
+package bigquery
```
[the remark this answers is not in the document]
Comment (kumare3) [reply go/tasks/plugins/webapi/bigquery/plugin.go:1]: shall we add a comment above the package name. I know this is not done everywhere, but we want to start doing this so that we get automated documentation

File: go/tasks/plugins/webapi/bigquery/plugin.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (kumare3) [reply go/tasks/plugins/webapi/bigquery/plugin.go:302]: we do have a method for non-retryable errors. Just the word `retryable` is dropeed - https://github.com/flyteorg/flyteplugins/blob/e7bf9a4e1b43ec7be55411fcc31babbe6c769db4/go/tasks/pluginmachinery/core/phase.go#L188

File: go/tasks/plugins/webapi/bigquery/plugin.go [hunk 22]
```diff
@@ -0,0 +1,512 @@
...
+	// transient errors are retryable, if any action is needed, errors are permanent
+
+	switch reason {
+	case "":
+		return pluginsCore.PhaseInfoSuccess(taskInfo)
+
+	// This error returns when you try to access a resource such as a dataset, table, view, or job that you
+	// don't have access to. This error also returns when you try to modify a read-only object.
+	case "accessDenied":
+		return pluginsCore.PhaseInfoFailed(pluginsCore.PhasePermanentFailure, userExecutionError, taskInfo)
```
Comment (kumare3) [inline go/tasks/plugins/webapi/bigquery/plugin.go:339]: https://github.com/flyteorg/flyteplugins/blob/e7bf9a4e1b43ec7be55411fcc31babbe6c769db4/go/tasks/pluginmachinery/core/phase.go#L188

File: go/tasks/plugins/webapi/bigquery/plugin.go [hunk 23]
```diff
@@ -0,0 +1,512 @@
...
+	if createError.Code >= http.StatusInternalServerError {
+		return core.PhaseInfoFailed(pluginsCore.PhaseRetryableFailure, systemExecutionError, taskInfo)
+	}
+
+	// something unexpected happened, just terminate task
+	return core.PhaseInfoFailed(pluginsCore.PhasePermanentFailure, systemExecutionError, taskInfo)
+}
+
+func handleErrorResult(reason string, message string, taskInfo *core.TaskInfo) core.PhaseInfo {
+	userExecutionError := &flyteIdlCore.ExecutionError{
```
Comment (kumare3) [inline go/tasks/plugins/webapi/bigquery/plugin.go:310]: also we have helper methods for system and user errors in the same file
https://github.com/flyteorg/flyteplugins/blob/e7bf9a4e1b43ec7be55411fcc31babbe6c769db4/go/tasks/pluginmachinery/core/phase.go#L188

**Review by kumare3 — COMMENTED** [review]
A few comments, more like nits
Just one major thing, the config for the plugin. Do we want that to be a protobuf? Or how is the java sdk sending this object today

**Comment by kanterov** [comment]
@kumare3 we have implemented dataclass for Python and auto-value class for Java/Scala. I can also write protobuf for it, but it isn't clear if it worth it. SDK code is already typed, and on wire it becomes proto struct.

File: go/tasks/plugins/webapi/bigquery/plugin.go (hunk 23 above)
Comment (kanterov) [reply go/tasks/plugins/webapi/bigquery/plugin.go:310]: Oh, nice! Thanks

File: boilerplate/flyte/golang_test_targets/Makefile (hunk 19 above)
Comment (kanterov) [reply boilerplate/flyte/golang_test_targets/Makefile]: I did that because linting was timing out. I think it's because of added dependencies. Let me try again.

**Comment by kanterov** [comment]
@kumare3 for protobuf, I started to write it, it will take some time, because there are many nested objects are imported from dependencies. 

One thing that I noticed is that currently go structs use camelCase for naming, while protobuf messages use snake_case. I'm not sure it's possible to change how protobuf serializes into structs. Underlying API will use camelCase, while proto convention is snake_case. Do you think we can use camelCase in proto?

**Comment by kumare3** [comment]
@kanterov that's ok, task template has versioning, so that you can easily choose the implementation.

I see that there is too much work. Let's go ahead with what you have. This might be a different example. 

I also have another idea, we can get the proto file from the go struct or use open api

**Review by kumare3 — APPROVED** [review]
